nemo/core/optim/adafactor.py

```diff
@@ -0,0 +1,216 @@
+# Copyright (c) 2021, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+# Most of the code here has been copied from:
+# https://github.com/pytorch/fairseq/blob/main/fairseq/optim/adafactor.py
+
+import math
+
+import torch
+from torch.optim.optimizer import Optimizer
+
+__all__ = ['Adafactor']
+
+
+class Adafactor(Optimizer):
```

Comment:
A naive question: do we rather implement our own optimizers rather than use externally implemented but tested ones?
How did we validate this implementation?

Reply:
Good point - I just trusted fairseq to implement Adafactor correctly. The only sanity check I did was to train a 1B param T5 with Adafactor and it converged fine.

Reply:
This should be fine for now. We should probably have a bigger discussion whether we want to maintain our own optimizers with the rest of the team. Currently we won't benefit from faiseq's bug fixes and extensions.